src/librustc_mir/transform/uniform_array_move_out.rs

```diff
@@ -0,0 +1,114 @@
+    }
+}
+
+impl UniformArrayMoveOut {
+    fn process_move_assign<'a, 'tcx>(&self,
+                                     tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                     mir: &Mir<'tcx>,
+                                     location: Location,
+                                     dst_place: &Place<'tcx>,
+                                     src_place: &Place<'tcx>,
+                                     patch: &mut MirPatch<'tcx>) {
+        if let Place::Projection(ref proj) = *src_place {
+            let place_ty = proj.base.ty(mir, tcx).to_ty(tcx);
+            if let ty::TyArray(_t, size_o) = place_ty.sty {
+                let opt_size = size_o.val.to_const_int().and_then(|v| v.to_u64());
+                if let Some(size) = opt_size {
+                    let size = size as u32;
+                    match proj.elem {
+                        ProjectionElem::Subslice{from, to} => {
+                            patch.make_nop(location);
+                            let to = size - to;
+                            let mut temps = vec![];
+                            for _i in from..to {
+                                let _temp = patch.new_temp(_t, mir.source_info(location).span);
+                                patch.add_statement( location, StatementKind::StorageLive(_temp));
+                                patch.add_assign(location,
+                                                 Place::Local(_temp),
+                                                 Rvalue::Use(
+                                                     Operand::Move(
+                                                         Place::Projection(box PlaceProjection{
+                                                             base: proj.base.clone(),
+                                                             elem: ProjectionElem::ConstantIndex{
+                                                                 offset: _i,
+                                                                 min_length: size,
+                                                                 from_end: false  }
+                                                         }))));
+                                temps.push(_temp);
+                            }
+                            patch.add_assign(location,
+                                             dst_place.clone(),
+                                             Rvalue::Aggregate(box AggregateKind::Array(_t),
+                                             temps.iter().map(
+                                                 |x| Operand::Move(Place::Local(*x))).collect()
+                                             ));
+                            for temp in temps {
+                                patch.add_statement(location, StatementKind::StorageDead(temp));
+                            }
+                        }
+                        ProjectionElem::ConstantIndex{offset, min_length: _, from_end: true} => {
```

Comment:
Some kind of comment would be good here. Probably:

---

Normalize "from-end" indices to operate from the start.